#Copyright 2012 Cumulus Networks, Inc.  All rights reserved.

Platform support for the DNI ET-7448 48x10+4x40 TOR platform

diff --git a/arch/powerpc/boot/dts/dni_7448.dts b/arch/powerpc/boot/dts/dni_7448.dts
new file mode 100644
index 0000000..f7a2563
--- /dev/null
+++ b/arch/powerpc/boot/dts/dni_7448.dts
@@ -0,0 +1,331 @@
+/*
+ * Delta Networks, Inc. ET-7448BF Device Tree Source
+ *
+ * Copyright 2012, Cumulus Networks, Inc.
+ *
+ * This program is free software; you can redistribute  it and/or modify it
+ * under  the terms of  the GNU General  Public License as published by the
+ * Free Software Foundation;  either version 2 of the  License, or (at your
+ * option) any later version.
+ *
+ */
+/dts-v1/;
+
+/ {
+	model = "dni,et-7448bf";
+	compatible = "dni,dni_7448";
+	#address-cells = <0x2>;
+	#size-cells = <0x2>;
+	aliases {
+		ethernet0 = &ENET2;
+		serial0 = &SERIAL0;
+		serial1 = &SERIAL1;
+		pci2    = &PCI2;
+	};
+	cpus {
+		#address-cells = <0x1>;
+		#size-cells = <0x0>;
+		PowerPC,P2020@0 {
+			device_type = "cpu";
+			reg = <0x0>;
+			next-level-cache = <0x1>;
+		};
+		PowerPC,P2020@1 {
+			device_type = "cpu";
+			reg = <0x1>;
+			next-level-cache = <0x1>;
+		};
+	};
+	memory {
+		device_type = "memory";
+	};
+	bcm_dma {
+		compatible = "early-dma-alloc";
+		// 64MB DMA region, aligned to 1MB
+		region_size = <0x04000000>;
+		alignment   = <0x00100000>;
+	};
+	mass_storage {
+		device = "mmcblk0";
+	};
+	localbus@ffe05000 {
+		#address-cells = <0x2>;
+		#size-cells = <0x1>;
+		compatible = "fsl,p2020-elbc", "fsl,elbc", "simple-bus";
+		reg = <0x0 0xffe05000 0x0 0x00001000>;
+		interrupts = <19 0x2>;
+		interrupt-parent = <&MPIC>;
+		ranges = <0x0 0x0 0x0 0xe8000000 0x8000000
+			  0x1 0x0 0x0 0xffdf0000 0x0000100>;
+		nor@0,0 {
+			#address-cells = <0x1>;
+			#size-cells = <0x1>;
+			compatible = "cfi-flash";
+			reg = <0x0 0x0 0x8000000>;
+			bank-width = <0x2>;
+			device-width = <0x2>;
+			byteswap;
+			partition@0 {
+				/* Entire flash minus (u-boot + onie) */
+				reg = <0x00000000 0x07b60000>;
+				label = "open";
+			};
+			partition@1 {
+				/* 4MB onie */
+				reg = <0x07b60000 0x00400000>;
+				label = "onie";
+			};
+			partition@2 {
+				/* 128KB, 1 sector */
+				reg = <0x07f60000 0x00020000>;
+				label = "uboot-env";
+				env_size = <0x2000>;
+			};
+			partition@3 {
+				/* 512KB u-boot */
+				reg = <0x07f80000 0x00080000>;
+				label = "uboot";
+			};
+		};
+		cpld@1,0 {
+			compatible = "dni,7448-cpld";
+			reg = <0x1 0x0 0x0000100>;
+		};
+	};
+	soc@ffe00000 {
+		#address-cells = <0x1>;
+		#size-cells = <0x1>;
+		device_type = "soc";
+		compatible = "fsl,p2020-immr", "simple-bus";
+		ranges = <0x0 0x0 0xffe00000 0x100000>;
+		bus-frequency = <0x0>;
+		memory-controller@2000 {
+			compatible = "fsl,p2020-memory-controller";
+			reg = <0x2000 0x1000>;
+			interrupt-parent = <&MPIC>;
+			interrupts = <0x12 0x2>;
+		};
+		I2C0: i2c@3000 {
+			#address-cells = <0x1>;
+			#size-cells = <0x0>;
+			cell-index = <0x0>;
+			compatible = "fsl-i2c";
+			reg = <0x3000 0x100>;
+			interrupts = <0x2b 0x2>;
+			interrupt-parent = <&MPIC>;
+			dfsrr;
+			rtc@68 {
+				compatible = "stm,m41st85";
+				reg = <0x68>;
+			};
+			board_eeprom@54 {
+				compatible = "at,24c08";
+				reg = <0x54>;
+				label = "board_eeprom";
+			};
+			spd@52 {
+				read-only;
+				compatible = "at,spd";
+				reg = <0x52>;
+			};
+		};
+		I2C1: i2c@3100 {
+			#address-cells = <0x1>;
+			#size-cells = <0x0>;
+			cell-index = <0x1>;
+			compatible = "fsl-i2c";
+			reg = <0x3100 0x100>;
+			interrupts = <0x2b 0x2>;
+			interrupt-parent = <&MPIC>;
+			dfsrr;
+			fan@1b {
+				compatible = "maxim,max6650";
+				reg = <0x1b>;
+			};
+			fan@1f {
+				compatible = "maxim,max6650";
+				reg = <0x1f>;
+			};
+			fan@48 {
+				compatible = "maxim,max6650";
+				reg = <0x48>;
+			};
+			fan@4b {
+				compatible = "maxim,max6650";
+				reg = <0x4b>;
+			};
+			hotswap@40 {
+				compatible = "lt,ltc4215";
+				reg = <0x40>;
+			};
+			hotswap@42 {
+				compatible = "lt,ltc4215";
+				reg = <0x42>;
+			};
+			tmon@49 {
+				compatible = "ti,tmp75";
+				reg = <0x49>;
+			};
+			tmon@4a {
+				compatible = "ti,tmp75";
+				reg = <0x4a>;
+			};
+			tmon@4c {
+				compatible = "ti,tmp75";
+				reg = <0x4c>;
+			};
+			tmon@4d {
+				compatible = "ti,tmp75";
+				reg = <0x4d>;
+			};
+			psu_eeprom@50 {
+				read-only;
+				label = "psu2_eeprom";
+				compatible = "at,24c08";
+				reg = <0x50>;
+			};
+			psu_eeprom@54 {
+				read-only;
+				label = "psu1_eeprom";
+				compatible = "at,24c08";
+				reg = <0x54>;
+			};
+		};
+		SERIAL0: serial@4500 {
+			cell-index = <0x0>;
+			device_type = "serial";
+			compatible = "ns16550";
+			reg = <0x4500 0x100>;
+			clock-frequency = <0x0>;
+			interrupts = <0x2a 0x2>;
+			interrupt-parent = <&MPIC>;
+		};
+		SERIAL1: serial@4600 {
+			cell-index = <0x1>;
+			device_type = "serial";
+			compatible = "ns16550";
+			reg = <0x4600 0x100>;
+			clock-frequency = <0x0>;
+			interrupts = <0x2a 0x2>;
+			interrupt-parent = <&MPIC>;
+		};
+		L2: l2-cache-controller@20000 {
+			compatible = "fsl,p2020-l2-cache-controller";
+			reg = <0x20000 0x1000>;
+			cache-line-size = <0x20>;
+			cache-size = <0x80000>;
+			interrupt-parent = <&MPIC>;
+			interrupts = <0x10 0x2>;
+		};
+		MDIO1: mdio@24520 {
+			#address-cells = <0x1>;
+			#size-cells = <0x0>;
+			compatible = "fsl,gianfar-mdio";
+			reg = <0x24520 0x20>;
+			PHY1: ethernet-phy@1 {
+				reg = <0x1>;
+				device_type = "ethernet-phy";
+			};
+			TBI1: tbi-phy@11 {
+				reg = <0x11>;
+				device_type = "tbi-phy";
+			};
+		};
+		ENET2: ethernet@26000 {
+			#address-cells = <0x1>;
+			#size-cells = <0x1>;
+			cell-index = <0x1>;
+			device_type = "network";
+			model = "eTSEC";
+			compatible = "gianfar";
+			reg = <0x26000 0x1000>;
+			ranges = <0x0 0x26000 0x1000>;
+			interrupts = <
+				   0x1f 0x2
+				   0x20 0x2
+				   0x21 0x2>;
+			interrupt-parent = <&MPIC>;
+			tbi-handle = <&TBI2>;
+			phy-handle = <&PHY1>;
+			phy-connection-type = "sgmii";
+		};
+		MDIO2: mdio@26520 {
+			#address-cells = <0x1>;
+			#size-cells = <0x0>;
+			compatible = "fsl,gianfar-tbi";
+			reg = <0x520 0x20>;
+			TBI2: tbi-phy@11 {
+				reg = <0x11>;
+				device_type = "tbi-phy";
+			};
+		};
+		SDHCI: sdhci@2e000 {
+			compatible = "fsl,p2020-esdhc", "fsl,esdhc";
+			reg = <0x2e000 0x1000>;
+			interrupts = <0x48 0x2>;
+			interrupt-parent = <&MPIC>;
+			clock-frequency = <0x0>;
+			// P2020 requires auto-cmd12
+			sdhci,auto-cmd12;
+		};
+		MPIC: pic@40000 {
+			interrupt-controller;
+			#address-cells = <0x0>;
+			#interrupt-cells = <0x2>;
+			reg = <0x40000 0x40000>;
+			compatible = "chrp,open-pic";
+			device_type = "open-pic";
+		};
+		msi@41600 {
+			compatible = "fsl,mpic-msi";
+			reg = <0x41600 0x80>;
+			msi-available-ranges = <0x0 0x100>;
+			interrupts = <
+				   0xe0 0x0
+				   0xe1 0x0
+				   0xe2 0x0
+				   0xe3 0x0
+				   0xe4 0x0
+				   0xe5 0x0
+				   0xe6 0x0
+				   0xe7 0x0>;
+			interrupt-parent = <&MPIC>;
+		};
+		global-utilities@e0000 {
+			compatible = "fsl,p2020-guts", "fsl,mpc8548-guts";
+			reg = <0xe0000 0x1000>;
+			fsl,has-rstcr;
+		};
+	};
+	PCI2: pcie@ffe0a000 {
+		compatible = "fsl,mpc8548-pcie";
+		device_type = "pci";
+		#interrupt-cells = <0x1>;
+		#size-cells = <0x2>;
+		#address-cells = <0x3>;
+		reg = <0x0 0xffe0a000 0x0 0x1000>;
+		bus-range = <0x0 0xff>;
+		ranges = <0x2000000 0x0 0xc0000000 0x0 0xc0000000 0x0 0x20000000
+			  0x1000000 0x0 0x00000000 0x0 0xffc20000 0x0 0x00010000>;
+		clock-frequency = <0x5f5e100>;
+		interrupt-parent = <&MPIC>;
+		interrupts = <0x1a 0x2>;
+		interrupt-map-mask = <0xf800 0x0 0x0 0x7>;
+		interrupt-map = <
+			      /* IDSEL 0x0 */
+			      0x0 0x0 0x0 0x1 &MPIC 0x0 0x1
+			      0x0 0x0 0x0 0x2 &MPIC 0x0 0x1
+			      0x0 0x0 0x0 0x3 &MPIC 0x0 0x1
+			      0x0 0x0 0x0 0x4 &MPIC 0x0 0x1>;
+
+
+		pcie@0 {
+			reg = <0x0 0x0 0x0 0x0 0x0>;
+			#size-cells = <0x2>;
+			#address-cells = <0x3>;
+			device_type = "pci";
+			ranges = <0x2000000 0x0 0xc0000000 0x2000000 0x0 0xc0000000 0x0 0x20000000
+				  0x1000000 0x0 0x00000000 0x1000000 0x0 0x00000000 0x0 0x10000>;
+		};
+	};
+};
diff --git a/arch/powerpc/platforms/85xx/dni_7448.c b/arch/powerpc/platforms/85xx/dni_7448.c
new file mode 100644
index 0000000..4618bef
--- /dev/null
+++ b/arch/powerpc/platforms/85xx/dni_7448.c
@@ -0,0 +1,261 @@
+/*
+ * DNI 7448 - P2020 setup and early boot code plus other random bits.
+ *
+ * Copyright 2012 Cumulus Networks, Inc.
+ *
+ * This program is free software; you can redistribute  it and/or modify it
+ * under  the terms of  the GNU General  Public License as published by the
+ * Free Software Foundation;  either version 2 of the  License, or (at your
+ * option) any later version.
+ */
+
+#include <linux/stddef.h>
+#include <linux/kernel.h>
+#include <linux/pci.h>
+#include <linux/kdev_t.h>
+#include <linux/delay.h>
+#include <linux/seq_file.h>
+#include <linux/interrupt.h>
+#include <linux/of_platform.h>
+#include <linux/early_dma_alloc.h>
+
+#include <asm/system.h>
+#include <asm/time.h>
+#include <asm/machdep.h>
+#include <asm/pci-bridge.h>
+#include <mm/mmu_decl.h>
+#include <asm/prom.h>
+#include <asm/udbg.h>
+#include <asm/mpic.h>
+
+#include <sysdev/fsl_soc.h>
+#include <sysdev/fsl_pci.h>
+
+#undef DEBUG
+
+#ifdef DEBUG
+#define DBG(fmt, args...) printk(KERN_ERR "%s: " fmt, __func__, ## args)
+#else
+#define DBG(fmt, args...)
+#endif
+
+/*
+ * Enough of the CPLD to reset the system... full driver loads as a module
+*/
+static uint8_t __iomem * cpld_regs;
+static uint32_t CPLD_RESET_REG = 0x20;
+
+/*******************************************************************************
+ *
+ * Platform initialization functions
+ *
+ *******************************************************************************
+*/
+
+/*
+ * Initialize the interrupt controller
+ */
+static void __init dni_7448_pic_init(void)
+{
+	struct mpic *mpic;
+	struct resource r;
+	struct device_node *np;
+
+	np = of_find_node_by_type(NULL, "open-pic");
+	if (np == NULL) {
+		printk(KERN_ERR "Could not find open-pic node\n");
+		return;
+	}
+
+	if (of_address_to_resource(np, 0, &r)) {
+		printk(KERN_ERR "Failed to map mpic register space\n");
+		of_node_put(np);
+		return;
+	}
+
+	mpic = mpic_alloc(np, r.start,
+		  MPIC_PRIMARY | MPIC_WANTS_RESET |
+		  MPIC_BIG_ENDIAN | MPIC_BROKEN_FRR_NIRQS |
+		  MPIC_SINGLE_DEST_CPU,
+		  0, 256, " OpenPIC  ");
+
+	BUG_ON(mpic == NULL);
+	of_node_put(np);
+
+	mpic_init(mpic);
+}
+
+/*
+ * Setup the architecture
+ */
+#ifdef CONFIG_SMP
+extern void __init mpc85xx_smp_init(void);
+#endif
+
+static void __init dni_7448_setup_arch(void)
+{
+	struct device_node *cpu;
+	const unsigned int *fp;
+#ifdef CONFIG_PCI
+	struct device_node *np;
+#endif
+	struct device_node *cpld;
+#ifdef CONFIG_MMC_SDHCI_OF_ESDHC
+	struct device_node *esdhc;
+	struct property    *clk_prop;
+	unsigned int       *clk;
+#endif
+
+	if (ppc_md.progress)
+		ppc_md.progress("dni_7448_setup_arch()", 0);
+
+	cpu = of_find_node_by_type(NULL, "cpu");
+	if (cpu != 0) {
+	     fp = of_get_property(cpu, "clock-frequency", NULL);
+	     if (fp != 0)
+		  loops_per_jiffy = *fp / HZ;
+	     else
+		  loops_per_jiffy = 500000000 / HZ;
+	     of_node_put(cpu);
+	}
+
+#ifdef CONFIG_PCI
+	for_each_node_by_type(np, "pci") {
+		if (of_device_is_compatible(np, "fsl,mpc8548-pcie"))
+			fsl_add_bridge(np, 0);
+	}
+#endif
+
+	cpld = of_find_compatible_node(NULL, NULL, "dni,7448-cpld");
+	if (!cpld) {
+		printk(KERN_ERR "Can not find 7448-cpld node in device tree\n");
+		cpld_regs = NULL;
+	} else {
+		cpld_regs = of_iomap(cpld, 0);
+		of_node_put(cpld);
+	}
+
+#ifdef CONFIG_MMC_SDHCI_OF_ESDHC
+	esdhc = of_find_compatible_node(NULL, NULL, "fsl,p2020-esdhc");
+	if (esdhc != NULL) {
+		clk_prop = of_find_property(esdhc, "clock-frequency", NULL);
+		if (clk_prop != NULL) {
+			/*
+			** Increase reported SDHC clock-frequency by 2x.
+			** This will cause the actual hardware bus speed
+			** to be 2x slower (calculated by driver).
+			**
+			** The 2x number is imperical.	One box I had
+			** required a 1.1x increase, while another one
+			** required a 1.4x increase.  Using 2x to
+			** hopefully never revisit.
+			**
+			** The slower hardware bus speed gives the part
+			** more margin for safe operation.  Without this
+			** fix-up the SDHC will sometimes report CRC
+			** errors on bus transactions and never recover.
+			*/
+			clk = clk_prop->value;
+			*clk = *clk * 2;
+		}
+		else {
+			printk(KERN_ERR "dni_7448: Can not find SDHC clock-frequency\n");
+		}
+		of_node_put(esdhc);
+	}
+	else {
+		printk(KERN_ERR "dni_7448: Can not find p2020-esdhc in device tree\n");
+	}
+#endif
+	powersave_nap = 0;
+
+#ifdef CONFIG_SMP
+	mpc85xx_smp_init();
+#endif
+
+#ifdef CONFIG_EARLY_DMA_ALLOC
+	eda_init();
+#endif
+}
+
+
+
+static struct of_device_id __initdata dni_7448_ids[] = {
+	{ .type = "soc", },
+	{ .compatible = "soc", },
+	{ .compatible = "simple-bus", },
+	{ .compatible = "gianfar", },
+	{},
+};
+
+
+
+static int __init dni_7448_publish_devices(void)
+{
+	return of_platform_bus_probe(NULL, dni_7448_ids, NULL);
+}
+machine_device_initcall(dni_7448, dni_7448_publish_devices);
+
+
+
+static void dni_7448_show_cpuinfo(struct seq_file *m)
+{
+	uint pvid, svid, phid1;
+	uint memsize = total_memory;
+
+	pvid = mfspr(SPRN_PVR);
+	svid = mfspr(SPRN_SVR);
+
+	seq_printf(m, "Vendor\t\t: Freescale Semiconductor\n");
+	seq_printf(m, "PVR\t\t: 0x%x\n", pvid);
+	seq_printf(m, "SVR\t\t: 0x%x\n", svid);
+
+	/* Display cpu Pll setting */
+	phid1 = mfspr(SPRN_HID1);
+	seq_printf(m, "PLL setting\t: 0x%x\n", ((phid1 >> 24) & 0x3f));
+
+	/* Display the amount of memory */
+	seq_printf(m, "Memory\t\t: %d MB\n", memsize / (1024 * 1024));
+}
+
+
+/*
+ * Platform specific restart... need to use the CPLD
+ */
+static void dni_7448_restart(char *cmd)
+{
+	printk (KERN_EMERG "Reset via the platform CPLD\n");
+
+	local_irq_disable();
+	writeb(0, (cpld_regs + CPLD_RESET_REG));
+	while (1);
+}
+
+/*
+ * Called very early, device-tree isn't unflattened
+ */
+static int __init dni_7448_probe(void)
+{
+	unsigned long root = of_get_flat_dt_root();
+
+	if (of_flat_dt_is_compatible(root, "dni,dni_7448"))
+		return 1;
+
+	return 0;
+}
+
+define_machine(dni_7448) {
+	.name		= "Delta Networks, Inc  ET-7448",
+	.probe		= dni_7448_probe,
+	.setup_arch	= dni_7448_setup_arch,
+	.init_IRQ	= dni_7448_pic_init,
+	.show_cpuinfo	= dni_7448_show_cpuinfo,
+#ifdef CONFIG_PCI
+	.pcibios_fixup_bus	= fsl_pcibios_fixup_bus,
+#endif
+	.get_irq	= mpic_get_irq,
+	.power_save	= e500_idle,
+	.restart	= dni_7448_restart,
+	.calibrate_decr = generic_calibrate_decr,
+	.progress	= udbg_progress,
+};
